### ISSUE TITLE
 Type 'React.ReactNode' is not assignable to type 'import("../node_modules/@types/react-transition-group/node_modules/@types/react/index").ReactNode'.  TS2786

### ISSUE THREAD

**Issue by asith-w** [formatjs/formatjs#3682]
## Which package?
^5.24.4

## Describe the bug
when npm build process is running below error 'Router' cannot be used as a JSX component. Its instance type 'BrowserRouter' is not a valid JSX element. is occurred. what could be the reason?

I am using node v14.18.3. package.json file is shown below

```
Creating an optimized production build...
  Failed to compile.
  
  D:/a/1/s/ClientApp/src/App.tsx
  TypeScript error in D:/a/1/s/ClientApp/src/App.tsx(20,6):
  'Router' cannot be used as a JSX component.
    Its instance type 'BrowserRouter' is not a valid JSX element.
      The types returned by 'render()' are incompatible between these types.
        Type 'React.ReactNode' is not assignable to type 'import("D:/a/1/s/ClientApp/node_modules/react-intl/node_modules/@types/react/index").ReactNode'.  TS2786
  
      18 | export default function App() {
      19 |   return (
    > 20 |     <Router>
         |      ^
      21 |       <Switch>
      22 |         <Route exact path="/"></Route>
      23 |         <Route path="/StaticTabs/MainTab" component={MainTab}></Route>
  
  
  npm ERR! code ELIFECYCLE
  npm ERR! errno 1
  npm ERR! abc-for-teams-by@0.1.0 build: `react-scripts build`
```

package.json

```
{
  "name": "abs",
  "version": "0.1.0",
  "private": true,
  "dependencies": {
    "@fluentui/react-northstar": "^0.63.0",    
    "@microsoft/teams-js": "^1.11.0",
    "@testing-library/jest-dom": "^5.16.1",
    "@testing-library/react": "^12.1.2",
    "@testing-library/user-event": "^13.5.0",
    "@types/microsoftteams": "^1.9.2", 
    "axios": "^0.21.1",
    "lodash": "^4.17.21",
    "msteams-react-base-component": "^1.1.1",
    "office-ui-fabric-react": "^7.181.0",
    "react": "^16.8.6",
    "react-dom": "^16.8.6",
    "react-hook-form": "^7.25.1",
    "react-intl": "^5.24.4",
    "react-router-dom": "^5.1.2",
    "react-scripts": "^4.0.3",
    "react-select": "^5.2.2",
    "read-appsettings-json": "^1.0.98",
    "web-vitals": "^2.1.4"
  },
  "devDependencies": {
    "@types/es6-promise": "^3.3.0",
    "@types/jest": "^27.4.0",
    "@types/lodash": "^4.14.182",
    "@types/node": "^16.11.21",
    "@types/react": "^16.8.6",
    "@types/react-dom": "^16.8.12",
    "@types/react-router-dom": "^5.1.7",
    "cross-env": "^7.0.3",
    "env-cmd": "^10.1.0",
    "typescript": "^4.5.5"
  },
  "scripts": {
    "start": "react-scripts start",
    "build": "react-scripts build",
    "test": "react-scripts test",
    "eject": "react-scripts eject"
  },
  "eslintConfig": {
    "extends": [
      "react-app",
      "react-app/jest"
    ],
    "rules": {
      "@typescript-eslint/no-array-constructor": "off",
      "array-callback-return": "off",
      "no-mixed-operators": "off",
      "no-empty-pattern": "off",
      "jsx-a11y/anchor-is-valid": "off"
    }
  },
  "browserslist": {
    "production": [
      ">0.2%",
      "not dead",
      "not op_mini all"
    ],
    "development": [
      "last 1 chrome version",
      "last 1 firefox version",
      "last 1 safari version"
    ]
  }
}
```

**Comment by longlho** [comment]
this looks like you have multiple versions of `react`

**Comment by TheBudgetMan** [comment]
We have the same problem when upgrading `react-intl` from 5.24.8 to 5.25.0. Suddenly we get a ton of compilation errors, some of them "simpler", like `'Container' cannot be used as a JSX component.
  Its instance type 'Container' is not a valid JSX element.` (for the native-base Container), some of them more complicated, like 
  
  ```
  The types returned by 'render()' are incompatible between these types.
      Type 'React.ReactNode' is not assignable to type 'import("...node_modules/react-intl/node_modules/@types/react/index").ReactNode'.
        Type '{}' is not assignable to type 'ReactNode'.
  ```

.babelrc
```
{
	"presets": [
		"module:metro-react-native-babel-preset",
		["@babel/preset-env", {"targets": {"node": "current"}}],
		"@babel/preset-typescript"
	],
	"sourceMaps": "inline",
	"plugins": [
		[
			"react-intl",
			{
				"messagesDir": ".tmp/messages/",
				"enforceDescriptions": false
			}
		],
		"react-native-reanimated/plugin"
	]
}
```

tsconfig.json
```
{
    "compilerOptions": {
        "target": "es6",
        "lib": [
            "es6"
        ],
        "module": "commonjs",
        "moduleResolution": "node",
        "outDir": "./lib",
        "jsx": "react",
        "experimentalDecorators": true,
        "emitDecoratorMetadata": true,
        "inlineSourceMap": true,
        "types": ["react", "react-native", "jest", "react-intl"],
        "strict": true,
        "skipLibCheck": true,
        "allowSyntheticDefaultImports": true,
        "strictPropertyInitialization": false,
        "resolveJsonModule": true,
    },
    "include": [
        "./src",
        "./themes",
        "node_modules/react-intl",
        "node_modules/intl-messageformat",
        "node_modules/@formatjs/icu-messageformat-parser"
    ],
    "exclude": ["node_modules"],
    "compileOnSave": true
}
```

React native version: 0.68.2
React version: 17.0.2

**Comment by longlho** [comment]
did you try removing your `node_modules` & reinstall all dependencies?

**Comment by TheBudgetMan** [comment]
> did you try removing your `node_modules` & reinstall all dependencies?

Yes, on every version switch.

**Comment by amelvinamica** [comment]
We are also on React 17.

If you look in your lock file (package-lock.json or yarn.lock), look at the react-intl definition and its dependencies. You'll likely see:

`"@types/react" "16 || 17 || 18"`

If you look up all the definitions for `@types/react`, there will be a definition for `"@types/react" "16 || 17 || 18"`. The version defined will be 18.x.x, thus react 18 types have been added into the project.

I'm not an expert on this stuff, so I would have thought the union between different type versions would have prevented the react 18 types being enforced. But that's what we experienced updating react-intl. It seems to update @types/react to a react 18 version, and it will include not just version 17 types but also version 18 types as well.

The fix we did was putting an explicit `@types/react` package version in the package.json under `devDependencies`.

Like this
```
"devDependencies": {
...
"@types/react": "17.0.44",
...
}
```

You will also need to make sure your lock file directs any package looking for `@types/react@16 || 17 || 18` to the react 17 types. So it will look like this:

```
"@types/react@*", "@types/react@17.0.44", "@types/react@^17", "@types/react@16 || 17 || 18":
  version "17.0.44"
...
```

When upgrading to React 18, that may need to be manually edited so react knows to install the react 18 types and use those instead.
